src/Juvix/FrontendContextualise/ModuleOpen/Environment.hs

```diff
@@ -100,3 +159,189 @@ lookupModMap s =
+    Just opens ->
+      -- TODO ∷
+      -- explicts and implicits for now work the same
+      put @"modMap" (foldr f modM opens)
+      where
+        f y modMap = foldr addNewSymbol modMap openList
+          where
+            nameSpace =
+              case y of
+                Implicit x -> x
+                Explicit x -> x
+            unpopulate = (() <$)
+            --
+            addNewSymbol x modMap =
+              case unpopulate (Context.lookup (NameSymbol.fromSymbol x) old)
+                <|> unpopulate (Context.lookup (NameSymbol.fromSymbol x) new) of
+                Just () ->
+                  -- if the symbol is in the map, then
+                  -- we don't shadow it....
+                  -- TODO ∷ later throw an error for
+                  --        explicit opens that do this
+                  modMap
+                Nothing ->
+                  Map.insert x nameSpace modMap
+            --
+            openList =
+              -- haskell gives type error if grabList is not typed
+              let grabList ::
+                    Context.T a b c ->
+                    NameSpace.List (Context.Definition a b c)
+                  grabList ctx =
+                    case Context.extractValue <$> Context.lookup nameSpace ctx of
+                      Just (Context.Record nameSpace _) ->
+                        NameSpace.toList nameSpace
+                      Just _ ->
+                        NameSpace.List [] []
+                      Nothing ->
+                        NameSpace.List [] []
+                  NameSpace.List {publicL = pubN} =
+                    grabList new
+                  NameSpace.List {publicL = pubO} =
+                    grabList old
+               in fmap fst pubN <> fmap fst pubO
+
+-- we don't take a module mapping as all symbols at the point of
+-- resolve can't be anything else, thus we don't have to pre-pend any
+-- qualification at this point... we add qualification later as we add
+-- things, as we have to figure out the full resolution of all opens
+
+resolve :: Context.T a b c -> [PreQualified] -> Either Error OpenMap
+resolve ctx = foldM (resolveSingle ctx) mempty
+
+resolveSingle ::
+  Context.T a b c -> OpenMap -> PreQualified -> Either Error OpenMap
+resolveSingle ctx openMap Pre {opens, implicitInner, explicitModule} =
+  case Context.switchNameSpace explicitModule ctx of
+    Left (Context.VariableShared err) ->
+      Left (IllegalModuleSwitch err)
+    Right ctx -> do
+      resolved <- pathsCanBeResolved ctx opens
+      qualifiedNames <- resolveLoop ctx mempty resolved
+      openMap
+        |> Map.insert explicitModule (fmap Explicit qualifiedNames)
+        |> ( \map ->
+               foldr
+                 (\mod -> Map.insert mod (fmap Implicit qualifiedNames))
+                 map
+                 implicitInner
+           )
+        |> Right
+
+-- this goes off after the checks have passed regarding if the paths
+-- are even possible to resolve
```

Comment:
add a comment describing pre, post-conditions, function behaviour

Reply:
Well this function deals with the single of the above, so any precondition would be from resolve... which doesn't really have any significant ones

Reply:
ditto, see above response